northstar_tests/build.rs

```diff
@@ -15,46 +15,67 @@
-    let bin = CargoBuild::new()
-        .manifest_path(CARGO_MANIFEST)
-        .current_release()
-        .target(env::var("TARGET").unwrap())
-        .target_dir("target")
-        .run()
-        .expect("Failed to build")
-        .path()
-        .to_owned();
-    fs::copy(&bin, &root.join("test_container")).expect("failed to copy bin");
+    for dir in &[
+        "../examples/cpueater",
+        "../examples/crashing",
+        "../examples/ferris",
+        "../examples/hello-ferris",
+        "../examples/hello-resource",
+        "../examples/hello-world",
+        "../examples/inspect",
+        "../examples/memeater",
+        "../examples/message-0.0.1",
+        "../examples/message-0.0.2",
+        "../examples/persistence",
+        "../examples/seccomp",
+        "test_container",
+        "test_resource",
+    ] {
+        let dir = Path::new(dir);
+        // Build crate if a Cargo manfiest is included in the directory
+        let cargo_manifest = dir.join("Cargo.toml");
 
-    let out_dir = env::var("OUT_DIR").unwrap();
+        let (root, tmpdir) = if cargo_manifest.exists() {
+            println!("Building {}", cargo_manifest.display());
+            let bin = CargoBuild::new()
+                .manifest_path(cargo_manifest)
+                .current_release()
+                .target(env::var("TARGET").unwrap())
+                .target_dir(Path::new("target").join("tests")) // Cannot reuse target because it's in use
+                .run()
+                .expect("Failed to build")
+                .path()
+                .to_owned();
 
-    // Pack test container npk
-    npk::npk::pack(
-        Path::new(TEST_CONTAINER_MANIFEST),
-        &npk.join("root"),
-        Path::new(&out_dir),
-        Some(Path::new(KEY)),
-    )
-    .expect("Failed to create test container npk");
+            println!("Binary is {}", bin.display());
+            let tmpdir = tempfile::TempDir::new().expect("Failed to create tmpdir");
+            let npk = tmpdir.path().join("npk");
+            let root = npk.join("root");
+            fs::create_dir_all(&root).expect("Failed to create npk root");
+            fs::copy(&bin, root.join(dir.file_name().unwrap())).expect("failed to copy bin");
+            (root, Some(tmpdir))
+        } else {
+            let root = dir.join("root");
+            if root.exists() {
+                (root, None)
+            } else {
+                (dir.to_owned(), None)
+            }
+        };
 
-    // Pack test resource npk
-    npk::npk::pack(
-        Path::new(TEST_RESOURCE_MANIFEST),
-        &Path::new("test_resource").join("root"),
-        Path::new(&out_dir),
-        Some(Path::new(KEY)),
-    )
-    .expect("Failed to create test resource npk");
+        npk::npk::pack(
+            &dir.join(Path::new("manifest.yaml")),
+            &root,
+            out_dir,
+            Some(Path::new(KEY)),
+        )
+        .expect("Failed to pack npk");
+        drop(tmpdir);
```

Comment:
Isn't `tmpdir` going out of scope her anyway?

Reply:
Sure. I just wanted to make it explicit.